### PR TITLE
Idempiere 4066

### PR DESCRIPTION
**Pull request by dpansheriya**
This enhancement adds RelativePeriodTo on report column. Using RelativePeriod and RelativePeriodTo, Range of periods (Inclusive) can specified.

If RelativePeriodTo is null, then it behaves as prior to this enhancement.
This enhancement are part of Adaxa Suite's Financial report improvements.  Test cases are mentioned on Ticket.

### CODE REVIEW

File: migration/i7.1z/postgresql/201912301730_IDEMPIERE-4066.sql [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+-- Oct 14, 2019 4:09:10 PM IST
+INSERT INTO AD_Field (AD_Field_ID,Name,Description,AD_Tab_ID,AD_Column_ID,IsDisplayed,DisplayLogic,DisplayLength,SeqNo,IsSameLine,IsHeading,IsFieldOnly,IsEncrypted,AD_Client_ID,AD_Org_ID,IsActive,Created,CreatedBy,Updated,UpdatedBy,IsReadOnly,IsCentrallyMaintained,EntityType,AD_Field_UU,IsDisplayedGrid,SeqNoGrid,XPosition,ColumnSpan,NumLines,IsQuickEntry,IsDefaultFocus,IsAdvancedField) VALUES (206213,'Relative Period To','Period offset (0 is current)',374,214087,'Y','@ColumnType@=R',26,145,'N','N','N','N',0,0,'Y',TO_TIMESTAMP('2019-10-14 16:09:10','YYYY-MM-DD HH24:MI:SS'),100,TO_TIMESTAMP('2019-10-14 16:09:10','YYYY-MM-DD HH24:MI:SS'),100,'N','Y','D','ebe78b9f-d030-46fa-b1c0-eaafe2521d3a','Y',145,4,2,1,'N','N','N')
+;
+
+-- Nov 26, 2019 4:04:43 PM IST
+UPDATE AD_Column SET DefaultValue=NULL,Updated=TO_TIMESTAMP('2019-11-26 16:04:43','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Column_ID=214087
+;
+
+-- Nov 26, 2019 4:04:46 PM IST
+INSERT INTO t_alter_column values('pa_reportcolumn','RelativePeriodTo','NUMERIC',null,'NULL')
```
Comment (globalqss) [inline migration/i7.1z/postgresql/201912301730_IDEMPIERE-4066.sql:23]: There is a mismatch between the oracle and postgres scripts - this line is not there, so maybe is better to review if is OK, or needs to be fixed, or not needed

File: org.adempiere.base/src/org/compiere/model/GridField.java [hunk 2]
```diff
@@ -156,6 +156,8 @@ protected void dispose()
 	/** Indicator for new Value				*/
 	public static final String  INSERTING = "FieldValueInserting";
 
+	private static final int	PA_REPORTCOLUMN_RelativePeriodTo_COLUMN_ID	= 214087;
```
Comment (globalqss) [inline org.adempiere.base/src/org/compiere/model/GridField.java:159]: This hardcode must go to the org.compiere.model.SystemIDs class

File: org.adempiere.base/src/org/compiere/report/MReportColumn.java [hunk 3]
```diff
@@ -442,6 +442,20 @@ public BigDecimal getRelativePeriod()
...
 		return null;
 	}	//	getRelativePeriod
+
+	/**
+	 * Get Relative Period To
+	 * 
+	 * @return relative period to
+	 */
+	@Override
+	public BigDecimal getRelativePeriodTo()
```
Comment (globalqss) [inline org.adempiere.base/src/org/compiere/report/MReportColumn.java:452]: when adding public methods - the serialVersionUID changes

**Review by globalqss — CHANGES_REQUESTED** [review]
I didn't test because it has compilation errors, please fix the requested and test.

**Comment by dpansheriya** [comment]
Ohk, Let me work on this fix.

File: migration/i7.1z/postgresql/201912301730_IDEMPIERE-4066.sql (hunk 1 above)
Comment (dpansheriya) [reply migration/i7.1z/postgresql/201912301730_IDEMPIERE-4066.sql:23]: This line is setting default value to null, while same is taken care in oracle script by 
`ALTER TABLE PA_ReportColumn ADD RelativePeriodTo NUMBER DEFAULT NULL
;`

**Comment by dpansheriya** [comment]
I committed changes on IDEMPIERE-4066 branch and looks pull request auto updated. This is great features. @globalqss  now you are good to test.

**Comment by globalqss** [comment]
@dpansheriya - you didn't answer - why is that code required in GridField?  what is special with the default of that column?

**Comment by dpansheriya** [comment]
> @dpansheriya - you didn't answer - why is that code required in GridField? what is special with the default of that column?

RelativePeriod_To is numeric. It was always getting defaulted to 0 but we want to use null as default. Null means not provided. 0 means selected current period on report process parameter. To avoid blank getting converted to 0, we added this.

**Comment by globalqss** [comment]
> > @dpansheriya - you didn't answer - why is that code required in GridField? what is special with the default of that column?
> 
> RelativePeriod_To is numeric. It was always getting defaulted to 0 but we want to use null as default. Null means not provided. 0 means selected current period on report process parameter. To avoid blank getting converted to 0, we added this.

OK, then that code is not required, you can define the default to NULL and the numeric fields will behave as expected.
That was added by @nmicoud in ticket IDEMPIERE-2678

**Comment by dpansheriya** [comment]
@globalqss Let me test this and if work then will revert.

**Comment by dpansheriya** [comment]
@globalqss 

I restored seed DB, applied corrected migration script on postgresql DB and tested below scenarios with reverted GridField code.
1. After Applying migration script, existing records in PA_ReportColumn have null value in relativePeriodTo (Verified in DB)
2. Set value relativePeriodTo = 5 and save
3. Making relativePeriodTo field blank and save. It reset value to null (Verified in DB)

**Review by globalqss — APPROVED** [review]

**Pull request merged**